libraries/botframework-connector/botframework/connector/auth/certificate_app_credentials.py

```diff
@@ -0,0 +1,57 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+from abc import ABC
+
+from msal import ConfidentialClientApplication
+
+from .app_credentials import AppCredentials
+
+
+class CertificateAppCredentials(AppCredentials, ABC):
+    """
+    AppCredentials implementation using a certificate.
+
+    See:
+    https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Client-Credentials#client-credentials-with-certificate
+    """
+
+    def __init__(
+        self,
+        app_id: str,
+        certificate_thumbprint: str,
+        certificate_private_key: str,
+        channel_auth_tenant: str = None,
+        oauth_scope: str = None,
+    ):
+        # super will set proper scope and endpoint.
+        super().__init__(
+            app_id=app_id,
+            channel_auth_tenant=channel_auth_tenant,
+            oauth_scope=oauth_scope,
+        )
+
+        self.scopes = [self.oauth_scope]
+        self.app = ConfidentialClientApplication(
+            client_id=self.microsoft_app_id,
+            authority=self.oauth_endpoint,
+            client_credential={
+                "thumbprint": certificate_thumbprint,
+                "private_key": certificate_private_key,
+            },
+        )
+
+    def get_token(self) -> str:
+        """
+        Implementation of AppCredentials.get_token.
+        :return: The access token for the given certificate.
+        """
+
+        # Firstly, looks up a token from cache
+        # Since we are looking for token for the current app, NOT for an end user,
+        # notice we give account parameter as None.
+        auth_token = self.app.acquire_token_silent(self.scopes, account=None)

```

Comment:
Did you inspect the code for acquire_token_for_client or try whether acquire_token_for_client caches? In the C# flavor of MSAL, acquire_token_for_client does the caching itself without external logic.

Reply:
It does appear to cache.  The first time through it falls into the fetch.  Next pass through it does not.  With that said, I did not test for expiration.